source/_posts/face-detection-with-dlib.md

```diff
@@ -0,0 +1,138 @@
+
+We'll learn how to detect faces in an image in few lines of code using python, [dlib](http://dlib.net/) and [opencv](https://opencv.org/).
+
+<!-- more -->
+
+## The Libraries
+
+[Dlib](http://dlib.net/) is a modern C++ toolkit containing machine learning algorithms and tools for creating complex software to solve real world problems. We are gonna use its [Python API](http://dlib.net/python/index.html).
+
+[OpenCV](https://opencv.org/) is a Computer Vision library also with interfaces in C++, Python and Java, and contains a lot of functions to help us to manipulate images.
+
+
+## Installing Dlib and Opencv
+
+Installing Python and its libraries is out of scope of this post. There are some tutorials out there that would help you get python, dlib and opencv installed. Here are a couple of links to install them on mac os:
+
+* https://www.learnopencv.com/install-dlib-on-macos/
+* https://www.learnopencv.com/install-opencv3-on-macos/
+
+## Face Detection
+
+In dlib's API we can find the `get_frontal_face_detector` method which "returns an object detector configured to detect human faces that are looking more or less towards the camera".
+
+The algorithm used to detect faces is [Histograms of Gradients (HOG)](https://lear.inrialpes.fr/people/triggs/pubs/Dalal-cvpr05.pdf) plus a [LinearSVM](https://en.wikipedia.org/wiki/Support_vector_machine).
+
+As I mentioned dlib also has a Neural Network implementation, it performs better than HOG to detect non-frontal faces, unfortunately it's computationally heavy and it's out of the scope in this post, if you are interested you could check the ["Sources"](#sources) section of this post for more info.
+
+In broad terms the HOG algorithm will iterate on every pixel of a given image, in each pixel it checks all the pixels around it and figures out how dark the current pixel is compared with the directly pixels surrounding it. Then it draws an arrow showing in which direction is getting darker. The process is repeated on every pixel, then it's break down in say 16x16 pixels squares where it counts up how many gradients point in each major direction, then that square is replaced with the arrow direction that was the strongest. Then using a lot of faces we could train a Linear Support Vector Machine model to detect faces in an image. - *Source:* [Machine Learning is Fun! Part 4: Modern Face Recognition with Deep Learning
+](https://medium.com/@ageitgey/machine-learning-is-fun-part-4-modern-face-recognition-with-deep-learning-c3cffc121d78)
+
+<img src="/face-detection-with-dlib/hog.png" />
+
+Dlib provides a the pretrained model using HOG through the `get_frontal_face_detector` method.
+
+## The Code
+
+We need to code the following steps:
+
+1. Import libraries.
+2. Get the dlib frontal face detector.
+3. Load the image using OpenCV.
+4. Pass de loaded image to the `detector`, it returns a array of rectangles, each rectangle containing the detected face bounds.
```

Comment:
"Pass de loaded the image to the `detector`; it returns **an** array of rectangles, each rectangle containing the detected face bounds."
is it "Pass  deloaded" or "Pass the loaded image"?